# Phase 1 Project

## Introduction

### Dataset explanation

For this project, I am presented with two datasets, "AviationData.csv" and "USState_Codes.csv," which were downloaded from Kaggle for analysis. Using this data, I am required to perform data cleaning and analysis to derive findings and actionable insights. These insights will help the head of a new aviation division make informed decisions about which aircraft to purchase. The analysis will focus on identifying patterns, trends, and key factors relevant to aircraft performance and safety.

## Objectives

## Description  of the datasets

### Importing Modules

In [59]:
##Importing pandas, matplotlib.pyplot, numpy, seaborn using aliases
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns

%matplotlib inline

### Loading Datasets

In [60]:
#Load AviationData.csv as a dataframe "df", change encoding, add memory
df = pd.read_csv('AviationData.csv', encoding='Latin', low_memory=False)
#Display the dataframe head
df.head()

Event.Id Investigation.Type Accident.Number  Event.Date  \
0  20001218X45444           Accident      SEA87LA080  1948-10-24   
1  20001218X45447           Accident      LAX94LA336  1962-07-19   
2  20061025X01555           Accident      NYC07LA005  1974-08-30   
3  20001218X45448           Accident      LAX96LA321  1977-06-19   
4  20041105X01764           Accident      CHI79FA064  1979-08-02   

          Location        Country   Latitude   Longitude Airport.Code  \
0  MOOSE CREEK, ID  United States        NaN         NaN          NaN   
1   BRIDGEPORT, CA  United States        NaN         NaN          NaN   
2    Saltville, VA  United States  36.922223  -81.878056          NaN   
3       EUREKA, CA  United States        NaN         NaN          NaN   
4       Canton, OH  United States        NaN         NaN          NaN   

  Airport.Name  ... Purpose.of.flight Air.carrier Total.Fatal.Injuries  \
0          NaN  ...          Personal         NaN                  2.0   
1          NaN  ...          Personal         NaN                  4.0   
2          NaN  ...          Personal         NaN                  3.0   
3          NaN  ...          Personal         NaN                  2.0   
4          NaN  ...          Personal         NaN                  1.0   

  Total.Serious.Injuries Total.Minor.Injuries Total.Uninjured  \
0                    0.0                  0.0             0.0   
1                    0.0                  0.0             0.0   
2                    NaN                  NaN             NaN   
3                    0.0                  0.0             0.0   
4                    2.0                  NaN             0.0   

  Weather.Condition  Broad.phase.of.flight   Report.Status Publication.Date  
0               UNK                 Cruise  Probable Cause              NaN  
1               UNK                Unknown  Probable Cause       19-09-1996  
2               IMC                 Cruise  Probable Cause       26-02-2007  
3               IMC                 Cruise  Probable Cause       12-09-2000  
4               VMC               Approach  Probable Cause       16-04-1980  

[5 rows x 31 columns]

### Information about the dataframe


In this analysis, I will explore the dataset structure using df.columns and determine its size with df.shape. I will also apply df.describe() to generate summary statistics, providing insights into key numeric values like fatalities and injuries, helping assess and prepare the data for further analysis.

In [61]:
#Find the columns in the dataframe
df.columns

Index(['Event.Id', 'Investigation.Type', 'Accident.Number', 'Event.Date',
       'Location', 'Country', 'Latitude', 'Longitude', 'Airport.Code',
       'Airport.Name', 'Injury.Severity', 'Aircraft.damage',
       'Aircraft.Category', 'Registration.Number', 'Make', 'Model',
       'Amateur.Built', 'Number.of.Engines', 'Engine.Type', 'FAR.Description',
       'Schedule', 'Purpose.of.flight', 'Air.carrier', 'Total.Fatal.Injuries',
       'Total.Serious.Injuries', 'Total.Minor.Injuries', 'Total.Uninjured',
       'Weather.Condition', 'Broad.phase.of.flight', 'Report.Status',
       'Publication.Date'],
      dtype='object')

In [62]:
#Finding the shape, number of columns and rows
df.shape

(88889, 31)

In [63]:
#Finding key statistics of the dataframe
df.describe

<bound method NDFrame.describe of              Event.Id Investigation.Type Accident.Number  Event.Date  \
0      20001218X45444           Accident      SEA87LA080  1948-10-24   
1      20001218X45447           Accident      LAX94LA336  1962-07-19   
2      20061025X01555           Accident      NYC07LA005  1974-08-30   
3      20001218X45448           Accident      LAX96LA321  1977-06-19   
4      20041105X01764           Accident      CHI79FA064  1979-08-02   
...               ...                ...             ...         ...   
88884  20221227106491           Accident      ERA23LA093  2022-12-26   
88885  20221227106494           Accident      ERA23LA095  2022-12-26   
88886  20221227106497           Accident      WPR23LA075  2022-12-26   
88887  20221227106498           Accident      WPR23LA076  2022-12-26   
88888  20221230106513           Accident      ERA23LA097  2022-12-29   

              Location        Country   Latitude   Longitude Airport.Code  \
0      MOOSE CREEK, ID  

In [64]:
#Finding the data types of columns in the dataset
df.dtypes

Event.Id                   object
Investigation.Type         object
Accident.Number            object
Event.Date                 object
Location                   object
Country                    object
Latitude                   object
Longitude                  object
Airport.Code               object
Airport.Name               object
Injury.Severity            object
Aircraft.damage            object
Aircraft.Category          object
Registration.Number        object
Make                       object
Model                      object
Amateur.Built              object
Number.of.Engines         float64
Engine.Type                object
FAR.Description            object
Schedule                   object
Purpose.of.flight          object
Air.carrier                object
Total.Fatal.Injuries      float64
Total.Serious.Injuries    float64
Total.Minor.Injuries      float64
Total.Uninjured           float64
Weather.Condition          object
Broad.phase.of.flight      object
Report.Status 

#### What have I learnt?

The dataset contains 88,889 aviation incidents with 31 columns detailing event dates, locations, aircraft info, and injury severities. It has missing values in key fields like geographic data and injuries, and mixed data types that require cleaning. The dataset contains different columns which have data of two data types, objects, and float values, 

## Cleaning the datasets

At this stage, null values and duplicates will be checked for. I intend to improve the quality and reliability of the dataset by removing duplicates and handling null values. This process ensures that redundant or incomplete data is eliminated, reducing potential biases and errors. Addressing missing data will be caried out either by removal or imputation, and eliminating duplicate records. 

### Finding with Null Values

In [65]:
# Calculating and displaying null values per column as a percentage of the entire DataFrame
null_percentage = (df.isnull().sum() / len(df)) * 100
print(null_percentage)


Event.Id                   0.000000
Investigation.Type         0.000000
Accident.Number            0.000000
Event.Date                 0.000000
Location                   0.058500
Country                    0.254250
Latitude                  61.320298
Longitude                 61.330423
Airport.Code              43.469946
Airport.Name              40.611324
Injury.Severity            1.124999
Aircraft.damage            3.593246
Aircraft.Category         63.677170
Registration.Number        1.481623
Make                       0.070875
Model                      0.103500
Amateur.Built              0.114750
Number.of.Engines          6.844491
Engine.Type                7.961615
FAR.Description           63.974170
Schedule                  85.845268
Purpose.of.flight          6.965991
Air.carrier               81.271023
Total.Fatal.Injuries      12.826109
Total.Serious.Injuries    14.073732
Total.Minor.Injuries      13.424608
Total.Uninjured            6.650992
Weather.Condition          5

In [66]:
#Finding duplicated rows in the dataset
df.duplicated().sum()

0

Based on a 20% missing data threshold from many data cleaning information sources, the columns that should be removed include Latitude (61.32%), Longitude (61.33%), Airport.Code (43.47%), Airport.Name (40.61%), Aircraft.Category (63.68%), FAR.Description (63.97%), Schedule (85.85%), and Air.carrier (81.27%). Additionally, columns like Total.Fatal.Injuries (12.83%), Total.Serious.Injuries (14.07%), Total.Minor.Injuries (13.42%), and Publication.Date (15.49%) may be considered for removal depending on their importance to the analysis. Removing these columns improves data quality by addressing significant gaps.

In [68]:
# Identifying columns with more than 20% missing data
missing_threshold = 0.20
columns_to_drop = df.columns[df.isnull().mean() > missing_threshold]

# Dropping the columns from the DataFrame
df = df.drop(columns=columns_to_drop)

# Display the columns that were dropped
print("Columns dropped due to >20% missing data:", columns_to_drop)

# Display the cleaned DataFrame (first few rows)
df.head()


Columns dropped due to >20% missing data: Index([], dtype='object')


Event.Id Investigation.Type Accident.Number  Event.Date  \
0  20001218X45444           Accident      SEA87LA080  1948-10-24   
1  20001218X45447           Accident      LAX94LA336  1962-07-19   
2  20061025X01555           Accident      NYC07LA005  1974-08-30   
3  20001218X45448           Accident      LAX96LA321  1977-06-19   
4  20041105X01764           Accident      CHI79FA064  1979-08-02   

          Location        Country Injury.Severity Aircraft.damage  \
0  MOOSE CREEK, ID  United States        Fatal(2)       Destroyed   
1   BRIDGEPORT, CA  United States        Fatal(4)       Destroyed   
2    Saltville, VA  United States        Fatal(3)       Destroyed   
3       EUREKA, CA  United States        Fatal(2)       Destroyed   
4       Canton, OH  United States        Fatal(1)       Destroyed   

  Registration.Number      Make  ... Number.of.Engines    Engine.Type  \
0              NC6404   Stinson  ...               1.0  Reciprocating   
1              N5069P     Piper  ...               1.0  Reciprocating   
2              N5142R    Cessna  ...               1.0  Reciprocating   
3              N1168J  Rockwell  ...               1.0  Reciprocating   
4               N15NY    Cessna  ...               NaN            NaN   

   Purpose.of.flight Total.Fatal.Injuries Total.Serious.Injuries  \
0           Personal                  2.0                    0.0   
1           Personal                  4.0                    0.0   
2           Personal                  3.0                    NaN   
3           Personal                  2.0                    0.0   
4           Personal                  1.0                    2.0   

   Total.Minor.Injuries  Total.Uninjured  Weather.Condition   Report.Status  \
0                   0.0              0.0                UNK  Probable Cause   
1                   0.0              0.0                UNK  Probable Cause   
2                   NaN              NaN                IMC  Probable Cause   
3                   0.0              0.0                IMC  Probable Cause   
4                   NaN              0.0                VMC  Probable Cause   

  Publication.Date  
0              NaN  
1       19-09-1996  
2       26-02-2007  
3       12-09-2000  
4       16-04-1980  

[5 rows x 22 columns]

I filled in missing values to clean the dataset. For numeric columns like Total.Fatal.Injuries and Total.Uninjured, I used the median to avoid the impact of extreme values. For categorical columns like Injury.Severity and Aircraft.damage, I used the most common value (mode). Additionally, for columns like Location and Country, I applied forward fill, which carries the last available value mode to fill any gaps. This ensures the dataset is complete and ready for analysis.

In [69]:
# Mean/Median imputation for numeric columns
df['Total.Fatal.Injuries'].fillna(df['Total.Fatal.Injuries'].median(), inplace=True)
df['Total.Serious.Injuries'].fillna(df['Total.Serious.Injuries'].median(), inplace=True)
df['Total.Minor.Injuries'].fillna(df['Total.Minor.Injuries'].median(), inplace=True)
df['Total.Uninjured'].fillna(df['Total.Uninjured'].median(), inplace=True)

# Mode imputation for categorical columns
df['Injury.Severity'].fillna(df['Injury.Severity'].mode()[0], inplace=True)
df['Aircraft.damage'].fillna(df['Aircraft.damage'].mode()[0], inplace=True)
df['Weather.Condition'].fillna(df['Weather.Condition'].mode()[0], inplace=True)
df['Registration.Number'].fillna(df['Registration.Number'].mode()[0], inplace=True)
df['Make'].fillna(df['Make'].mode()[0], inplace=True)
df['Model'].fillna(df['Model'].mode()[0], inplace=True)
df['Amateur.Built'].fillna(df['Amateur.Built'].mode()[0], inplace=True)
df['Number.of.Engines'].fillna(df['Number.of.Engines'].mode()[0], inplace=True)
df['Engine.Type'].fillna(df['Engine.Type'].mode()[0], inplace=True)
df['Purpose.of.flight'].fillna(df['Purpose.of.flight'].mode()[0], inplace=True)
df['Report.Status'].fillna(df['Report.Status'].mode()[0], inplace=True)
df['Publication.Date'].fillna(df['Publication.Date'].mode()[0], inplace=True)

# Mode imputation for 'Location' and 'Country' as aviation incidents are independent
df['Location'].fillna(df['Location'].mode()[0], inplace=True)
df['Country'].fillna(df['Country'].mode()[0], inplace=True)

In [70]:
df.isnull().values.any()

False

In [47]:
df

Event.Id Investigation.Type Accident.Number  Event.Date  \
0      20001218X45444           Accident      SEA87LA080  1948-10-24   
1      20001218X45447           Accident      LAX94LA336  1962-07-19   
2      20061025X01555           Accident      NYC07LA005  1974-08-30   
3      20001218X45448           Accident      LAX96LA321  1977-06-19   
4      20041105X01764           Accident      CHI79FA064  1979-08-02   
...               ...                ...             ...         ...   
88884  20221227106491           Accident      ERA23LA093  2022-12-26   
88885  20221227106494           Accident      ERA23LA095  2022-12-26   
88886  20221227106497           Accident      WPR23LA075  2022-12-26   
88887  20221227106498           Accident      WPR23LA076  2022-12-26   
88888  20221230106513           Accident      ERA23LA097  2022-12-29   

              Location        Country Injury.Severity Aircraft.damage  \
0      MOOSE CREEK, ID  United States        Fatal(2)       Destroyed   
1       BRIDGEPORT, CA  United States        Fatal(4)       Destroyed   
2        Saltville, VA  United States        Fatal(3)       Destroyed   
3           EUREKA, CA  United States        Fatal(2)       Destroyed   
4           Canton, OH  United States        Fatal(1)       Destroyed   
...                ...            ...             ...             ...   
88884    Annapolis, MD  United States           Minor     Substantial   
88885      Hampton, NH  United States       Non-Fatal     Substantial   
88886       Payson, AZ  United States       Non-Fatal     Substantial   
88887       Morgan, UT  United States       Non-Fatal     Substantial   
88888       Athens, GA  United States           Minor     Substantial   

      Registration.Number                        Make  ... Number.of.Engines  \
0                  NC6404                     Stinson  ...               1.0   
1                  N5069P                       Piper  ...               1.0   
2                  N5142R                      Cessna  ...               1.0   
3                  N1168J                    Rockwell  ...               1.0   
4                   N15NY                      Cessna  ...               1.0   
...                   ...                         ...  ...               ...   
88884              N1867H                       PIPER  ...               1.0   
88885              N2895Z                    BELLANCA  ...               1.0   
88886              N749PJ  AMERICAN CHAMPION AIRCRAFT  ...               1.0   
88887              N210CU                      CESSNA  ...               1.0   
88888              N9026P                       PIPER  ...               1.0   

         Engine.Type  Purpose.of.flight Total.Fatal.Injuries  \
0      Reciprocating           Personal                  2.0   
1      Reciprocating           Personal                  4.0   
2      Reciprocating           Personal                  3.0   
3      Reciprocating           Personal                  2.0   
4      Reciprocating           Personal                  1.0   
...              ...                ...                  ...   
88884  Reciprocating           Personal                  0.0   
88885  Reciprocating           Personal                  0.0   
88886  Reciprocating           Personal                  0.0   
88887  Reciprocating           Personal                  0.0   
88888  Reciprocating           Personal                  0.0   

      Total.Serious.Injuries  Total.Minor.Injuries  Total.Uninjured  \
0                        0.0                   0.0              0.0   
1                        0.0                   0.0              0.0   
2                        0.0                   0.0              1.0   
3                        0.0                   0.0              0.0   
4                        2.0                   0.0              0.0   
...                      ...                   ...              ...   
88884                    1.0                